model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java

```diff
@@ -0,0 +1,273 @@
+    @Basic(fetch = FetchType.LAZY)
+    private String name;
+
+    @OneToMany(mappedBy = "role", cascade = CascadeType.PERSIST, orphanRemoval = true)
+    private final Set<JpaRoleAttributeEntity> attributes = new HashSet<>();
+
+    /**
+     * No-argument constructor, used by hibernate to instantiate entities.
+     */
+    public JpaRoleEntity() {
+        this.metadata = new JpaRoleMetadata();
+    }
+
+    public JpaRoleEntity(DeepCloner cloner) {
+        this.metadata = new JpaRoleMetadata(cloner);
+    }
+
+    /**
+     * Used by hibernate when calling cb.construct from read(QueryParameters) method.
+     * It is used to select role without metadata(json) field.
+     */
+    public JpaRoleEntity(UUID id, Integer entityVersion, String name) {
+        this.id = id;
+        this.entityVersion = entityVersion;
+        this.name = name;
+        this.metadata = null;
+    }
+
+    public boolean isMetadataInitialized() {
+        return metadata != null;
+    }
+
+    /**
+     * In case of any update on entity, we want to update the entityVerion
+     * to current one.
+     */
+    private void checkEntityVersionForUpdate() {
+        Integer ev = getEntityVersion();
+        if (ev != null && ev < SUPPORTED_VERSION_ROLE) {
+            setEntityVersion(SUPPORTED_VERSION_ROLE);
+        }
+    }
+
+    public Integer getEntityVersion() {
+        if (isMetadataInitialized()) return metadata.getEntityVersion();
+        return entityVersion;
+    }
+
+    public void setEntityVersion(Integer entityVersion) {
+        metadata.setEntityVersion(entityVersion);
+    }
+
+    @Override
+    public String getId() {
+        return id == null ? null : id.toString();
+    }
+
+    @Override
+    public void setId(String id) {
+        this.id = id == null ? null : UUID.fromString(id);
+    }
+
+    @Override
+    public Boolean isClientRole() {
+        return metadata.isClientRole();
+    }
+
+    @Override
+    public String getRealmId() {
+        return metadata.getRealmId();
+    }
+
+    @Override
+    public String getClientId() {
+        return metadata.getClientId();
+    }
+
+    @Override
+    public String getName() {
+        if (isMetadataInitialized()) return metadata.getName();
+        return name;
+    }
+
+    @Override
+    public String getDescription() {
+        return metadata.getDescription();
+    }
+
+    @Override
+    public void setClientRole(Boolean clientRole) {
+        checkEntityVersionForUpdate();
+        metadata.setClientRole(clientRole);
+    }
+
+    @Override
+    public void setRealmId(String realmId) {
+        checkEntityVersionForUpdate();
+        metadata.setRealmId(realmId);
+    }
+
+    @Override
+    public void setClientId(String clientId) {
+        checkEntityVersionForUpdate();
+        metadata.setClientId(clientId);
+    }
+
+    @Override
+    public void setName(String name) {
+        checkEntityVersionForUpdate();
+        metadata.setName(name);
```

Comment:
I suppose this should set the searchable field name on the entity as well by adding

```
this.name = name;
```

Reply:
This field is marked as "non-insertable" and "non-updatable": https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java#L63

It is generated from json and it's happening on database level, so when update happens it's updated only in metadata object and then database generates value of this field automatically. It's used when we want to load just subset of fields using `read(QueryParameters<RoleModel> queryParameters)`: https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleMapKeycloakTransaction.java#L77-L81

Reply:
Thanks, this explains it for me. As this was "surprising" to me, I'd be happy to see 2-3 comments in the code what's happening there / why the annotations have been placed here and what is happening in the background, as it might help future developers/maintainers of the code. Just not too much comments in too many places to risk that they run out of sync.

Reply:
I've added a javadoc to both `JpaRoleEntity` and `JpaClientEntity` where I've tried to describe the behavior.

Reply:
Looks good to me, thanks!